### PR TITLE
DSDEGP-2062: cram, arrays delivery refinements

### PR DESCRIPTION
**Pull request by pshapiro4broad**
### Description

https://broadinstitute.atlassian.net/browse/DSDEGP-2062

- refactored common delivery code into `DeliverExecutor`
- added check for existing workspace name in metadata to prevent re-delivery of already delivered data

----

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

### CODE REVIEW

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    moveOps: immutable.Seq[IoOp]
+  ): Source[(moveCommand.index.MetadataType, immutable.Seq[IoOp]), NotUsed]
+
+  override def checkPreconditions(
+    ioUtil: IoUtil,
+    webClient: ClioWebClient
+  ): Source[moveCommand.index.MetadataType, NotUsed] = {
+    val baseStream = super.checkPreconditions(ioUtil, webClient)
+    baseStream.flatMapConcat {
+      case (metadata) =>
```
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: `case` and parens not needed here.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        if (metadata.workspaceName.isDefined && !metadata.workspaceName.get.equals(
```
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: I think this is `metadata.workspaceName.exists(_ != deliverCommand.workspaceName)`

File: clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala [hunk 3]
```diff
@@ -1172,4 +1172,179 @@ trait WgsCramTests { self: BaseIntegrationSpec =>
       storedDocument1.unsafeGet[String]("notes") should be("I'm a note")
     }
   }
+
+  it should "not allow a delivery if a record has already been delivered" in {
```
Comment (danxmoran) [inline clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala]: What problems did you hit trying to do this in the unit tests?

Comment (pshapiro4broad) [reply clio-integration-test/src/it/scala/org/broadinstitute/clio/integrationtest/tests/WgsCramTests.scala]: I was able to create a non-sealed base class for all commands, but it was hard to create the code to execute my test only command using the client main code. I have it shelved in git and I'll make another attempt on it tomorrow.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala (hunk 2 above)
Comment (pshapiro4broad) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: Yeah, I had something a lot like that before but was still wrestling types at the time. I'll change it.

**Review by als364 — APPROVED** [review]

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  ): Source[(moveCommand.index.MetadataType, immutable.Seq[IoOp]), NotUsed]
+
+  override def checkPreconditions(
+    ioUtil: IoUtil,
+    webClient: ClioWebClient
+  ): Source[moveCommand.index.MetadataType, NotUsed] = {
+    val baseStream = super.checkPreconditions(ioUtil, webClient)
+    baseStream.flatMapConcat { metadata =>
+      if (metadata.workspaceName.exists(
+            _ != deliverCommand.workspaceName && !deliverCommand.force
```
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: Would bust the `deliverCommand.force` out of & before the `exists`:
```scala
if (!deliverCommand.force && metadata.workspaceName.exists(...))
```

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala [hunk 5]
```diff
@@ -20,68 +19,60 @@ import scala.concurrent.ExecutionContext
...
-              }
-
-            val newMetadata = movedMetadata
-              .withWorkspaceName(deliverCommand.workspaceName)
-              .copy(
-                grnIdatPath = Some(movedGrnIdat),
-                redIdatPath = Some(movedRedIdat)
-              )
+        val newMetadata = movedMetadata
+          .withWorkspaceName(deliverCommand.workspaceName)
```
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: Could we move the call to `withWorkspaceName` into the superclass?

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverWgsCramExecutor.scala [hunk 6]
```diff
@@ -22,37 +21,29 @@ import scala.concurrent.ExecutionContext
...
+
+  override protected def buildDelivery(
+    movedMetadata: WgsCramMetadata,
+    moveOps: immutable.Seq[IoOp]
+  ): Source[(moveCommand.index.MetadataType, immutable.Seq[IoOp]), NotUsed] = {
+    (movedMetadata.cramMd5, movedMetadata.cramPath) match {
+      case (Some(cramMd5), Some(cramPath)) => {
+
+        val newMetadata =
+          movedMetadata.withWorkspaceName(deliverCommand.workspaceName)
```
Comment (danxmoran) [inline clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverWgsCramExecutor.scala]: Same question about moving this to the superclass

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+          force = force
+        )
+      )
+    executor.execute(webClient, ioUtil).runWith(Sink.head).map { json =>
+      json.as[UpsertId] should be(Right(id))
+    }
+  }
+
+  it should "deliver wgs cram with base name change" in {
+    successfulDelivery()
```
Comment (danxmoran) [inline clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala]: I think this is technically a feature of the move executor.

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        WgsCramIndex,
+        theKey,
+        metadata.copy(
+          cramPath = Some(destinationCramPath),
+          craiPath = Some(destinationCraiPath),
+          workspaceName = Some(workspaceName)
+        ),
+        true
+      )
+      .returning(Source.single(id.asJson))
```
Comment (danxmoran) [inline clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala]: Do we need to duplicate most of this setup from the move executor, or can we just test the overridden methods?

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  private def successfulDelivery(
+    metadata: WgsCramMetadata = metadata,
+    force: Boolean = false
+  ) = {
+    val (ioUtil: IoUtil, webClient: ClioWebClient) = initMocks(metadata)
+
+    val executor =
+      new DeliverWgsCramExecutor(
```
Comment (danxmoran) [inline clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala]: Using `DeliverWgsCramExecutor` brings along the baggage of the extra cram operations it defines. If we want to purely test the superclass, I'd make an anonymous instance with the abstract method overridden to be a no-op.

File: clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala (hunk 7 above)
Comment (pshapiro4broad) [reply clio-client/src/test/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutorSpec.scala]: Sure, considering that the trivial delivery is a move.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala (hunk 4 above)
Comment (pshapiro4broad) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: It depends on if you want to short-circuit the workspace name check or the force check. Doesn't matter to me either way.

Comment (danxmoran) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: I think it'll just make for better formatting 😄 but the pattern we use elsewhere is for `force` to short-circuit even performing a check.

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala (hunk 5 above)
Comment (pshapiro4broad) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverArraysExecutor.scala]: OK

File: clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala (hunk 4 above)
Comment (pshapiro4broad) [reply clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/DeliverExecutor.scala]: No such luck regarding the formatting
```
      if (deliverCommand.force || !metadata.workspaceName.exists(
            _ != deliverCommand.workspaceName
          )) {
```

**Review by danxmoran — APPROVED** [review]

**Pull request merged**